blocks/file-blocks/markdown-edit/copy-widget.tsx

```diff
@@ -0,0 +1,316 @@
+        if (type.name.startsWith("ATXHeading")) {
+          const level = type.name.split("Heading")[1];
+          const newDecoration = headerDecoration({ level });
+          widgets.push(
+            newDecoration.range(
+              state.doc.lineAt(from).from,
+              state.doc.lineAt(to).to
+            )
+          );
+        } else if (type.name === "SetextHeading2") {
+          const newDecoration = horizontalRuleDecorationAfter();
+          widgets.push(newDecoration.range(from, to));
+        } else if (type.name === "Link") {
+          const text = state.doc.sliceString(from, to);
+          const linkRegex = /\[.*?\]\((?<url>.*?)\)/;
+          const url = linkRegex.exec(text)?.groups?.url;
+          const linkTextRegex = /\[(?<text>.*?)\]/;
+          const linkText = linkTextRegex.exec(text)?.groups?.text || "";
+          if (url) {
+            const newDecoration = linkAltDecoration(text, linkText, url);
+            widgets.push(newDecoration.range(from));
+            const newAltDecoration = linkDecoration(text, linkText, url);
+            widgets.push(newAltDecoration.range(from, to));
+          }
+        } else if (type.name === "Blockquote") {
+          const newDecoration = blockquoteDecoration();
+          widgets.push(newDecoration.range(from));
+        } else if (type.name === "ListItem") {
+          const text = state.doc.sliceString(from, to);
+          const listType = ["-", "*"].includes(text[0]) ? "ul" : "ol";
+          const index = text.split(" ")[0];
+          const newDecoration = listItemDecoration(listType, index);
+          widgets.push(newDecoration.range(from));
+        } else if (type.name === "CodeText") {
+          const newDecoration = codeBlockDecoration();
+          const fromLine = state.doc.lineAt(from);
+          const toLine = state.doc.lineAt(to);
+          for (let i = fromLine.from; i < toLine.to; i++) {
+            const linePosition = state.doc.lineAt(i);
+            widgets.push(newDecoration.range(linePosition.from));
+          }
+          // widgets.push(newDecoration.range(from));
+        } else if (type.name === "HTMLTag") {
+          // punting for now, it splits the text into multiple widgets:
+          // start tag, text, end tag
+          // const linkRegexHtml = /<a.*?href="(?<url>.*?)".*?>(?<text>.*?)[<\/a>]*/
+          // const result = linkRegexHtml.exec(state.doc.sliceString(from, to))
+          // if (result && result.groups && result.groups.url) {
+          //   let linkText = result.groups.text
+          //   if (!linkText && !linkText.includes("</a>")) {
+          //     const nextNode = tree.resolve(to)
+          //     linkText = state.doc.slice(nextNode.from, nextNode.to).text[0]
+          //     to += linkText.length
+          //   }
+          //   const url = result.groups.url
+          //   const newDecoration = linkDecoration(linkText, url)
+          //   widgets.push(newDecoration.range(from, to))
+          // }
```

Comment:
let's nix this for clarity!

Reply:
I'll do you one better and just implement it!